lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -793,6 +741,90 @@ def __init__(self, charm: CharmBase, relationship_name: str):
+        provider_relation_data = _load_relation_data(relation.data[self.model.app])
+        return provider_relation_data.get("certificates", [])
+
+    def _requirer_csrs(self, unit) -> List[Dict[str, str]]:
+        """Returns list of requirer CSR's from relation data."""
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            raise RuntimeError(f"Relation {self.relationship_name} does not exist")
+        requirer_relation_data = _load_relation_data(relation.data[unit])
+        return requirer_relation_data.get("certificate_signing_requests", [])
+
+    def _add_certificate(
+        self, certificate: str, certificate_signing_request: str, ca: str, chain: str
+    ) -> None:
+        """Adds certificate to relation data.
+
+        Args:
+            certificate (str): Certificate
+            certificate_signing_request (str): Certificate Signing Request
+            ca (str): CA Certificate
+            chain (str): CA Chain
+
+        Returns:
+            None
+        """
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            raise RuntimeError(
+                f"Relation {self.relationship_name} does not exist - "
+                f"The certificate request can't be completed"
+            )
+        new_certificate = {
+            "certificate": certificate,
+            "certificate_signing_request": certificate_signing_request,
+            "ca": ca,
+            "chain": chain,
+        }
+        certificates = copy.deepcopy(self._provider_certificates)
+        if new_certificate in certificates:
+            logger.info("Certificate already in relation data - Doing nothing")
+            return
+        certificates.append(new_certificate)
+        relation.data[self.model.app]["certificates"] = json.dumps(certificates)
+
+    def _remove_certificate(
+        self,
+        relation_id: int,
+        certificate: str = None,
+        certificate_signing_request: str = None,
+    ) -> None:
+        """Removes certificate from a given relation based on user provided certificate or csr.
+
+        Args:
+            relation_id (int): Relation id
+            certificate (str): Certificate (optional)
+            certificate_signing_request: Certificate signing request (optional)
+
+        Returns:
+            None
+        """
+        relation = self.model.get_relation(
+            relation_name=self.relationship_name,
+            relation_id=relation_id,
+        )
+        if not relation:
+            raise RuntimeError(
+                f"Relation {self.relationship_name} with relation id {relation_id} does not exist"
+            )
+        certificates = copy.deepcopy(self._provider_certificates)
+        for certificate_dict in certificates:
+            if certificate:
+                if certificate_dict["certificate"] == certificate:
```

Comment:
Is there any reason why we don't merge both conditions?
```suggestion
           if certificate and certificate_dict["certificate"] == certificate:
```

Reply:
Done